geopandas/clip.py

```diff
@@ -0,0 +1,228 @@
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a clipped subset of shp
+        containing multi-point and point features.
+    """
+
+    # Explode multi-point features when clipping then recreate geom
+    clipped = _clip_points(shp.explode().reset_index(level=[1]), clip_obj)
+    clipped = clipped.dissolve(by=[clipped.index]).drop(columns="level_1")[
+        shp.columns.tolist()
+    ]
+
+    return clipped
+
+
+def _clip_line_poly(shp, clip_obj):
+    """Clip line and polygon geometry to the clip_obj GeoDataFrame extent.
+
+    Clip an input line or polygon to the polygon extent of the clip_obj
+    parameter. Lines or Polygons that intersect the clip_obj geometry are
+    extracted with associated attributes and returned.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+        Line or polygon geometry that is clipped to clip_obj.
+
+    clip_obj : GeoDataFrame
+        Reference polygon for clipping.
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a clipped subset of shp
+        that intersects with clip_obj.
+    """
+    # Create a single polygon object for clipping
+    poly = clip_obj.geometry.unary_union
+    spatial_index = shp.sindex
+
+    # Create a box for the initial intersection
+    bbox = poly.bounds
+    # Get a list of id's for each object that overlaps the bounding box and
+    # subset the data to just those lines
+    sidx = list(spatial_index.intersection(bbox))
+    shp_sub = shp.iloc[sidx]
+
+    # Clip the data - with these data
+    clipped = shp_sub.copy()
+    clipped["geometry"] = shp_sub.intersection(poly)
+
+    # Return the clipped layer with no null geometry values
+    return clipped[clipped.geometry.notnull()]
+
+
+def _clip_multi_poly_line(shp, clip_obj):
+    """Clip multi lines and polygons to the clip_obj GeoDataFrame extent.
+
+    Clip an input multi line or polygon to the polygon extent of the clip_obj
+    parameter. Lines or Polygons that intersect the clip_obj geometry are
+    extracted with associated attributes and returned.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+        multiLine or multipolygon geometry that is clipped to clip_obj.
+
+    clip_obj : GeoDataFrame
+        Reference polygon for clipping.
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a clipped subset of shp
+        that intersects with clip_obj.
+    """
+
+    # Clip multi polygons
+    clipped = _clip_line_poly(shp.explode().reset_index(level=[1]), clip_obj)
+
+    lines = clipped[
+        (clipped.geometry.type == "MultiLineString")
+        | (clipped.geometry.type == "LineString")
+    ]
+    line_diss = lines.dissolve(by=[lines.index]).drop(columns="level_1")
+
+    polys = clipped[clipped.geometry.type == "Polygon"]
+    poly_diss = polys.dissolve(by=[polys.index]).drop(columns="level_1")
+
+    return gpd.GeoDataFrame(pd.concat([poly_diss, line_diss], ignore_index=True))
```

Comment:
Some questions about the implementation here (it might that adding some comments explaining this is enough):

- Similar to my question above, but is there a reason to explode multi-objects first? I would suppose that `intersection` works fine on those as well, or does that give a different result?
- Is there a reason that you need to dissolve the the linestrings and polygons separately?

Reply:
@jorisvandenbossche sure we struggled with this and it's a bit messy.

i'll revisit the code a bit more after i go through the rest of your comments.

this mess represents.
1. an edge case where sometimes you are clipping a polygon gdf and you get a sliver back (ie lines AND polygons). this is not the expected return but it can happen. in that case, you need to grab those slivers, reassociate the id and associated attributes (tabular data) and put the geometry back together. similarly for polygons. i was struggling with writing code that worked for both as one time. 

let me revisit this to see if i can add better documentation (comments) OR perhaps consolidate the code a bit.

We did chose to use intersects rather than intersection for performance reasons. let me test this a bit more.

Reply:
Hey @jorisvandenbossche ! I'm helping @lwasser with this PR, and did some digging into why we explode multi-objects first, and hope I can explain in a bit more detail. I made a gist about the issues we were seeing and why we have to handle multi-objects differently. If this doesn't make sense let me know and I can go into further detail! https://gist.github.com/nkorinek/92a9fbdc6a8a83909bb6bc51d1d94d70